libs/ngxtension/computed-async/src/computed-async.ts

```diff
@@ -0,0 +1,158 @@
+ * );
+ * ```
+ *
+ * The computed value will be `null` until the promise resolves.
+ * Everytime the userId changes, the fetch will be called again, and the previous fetch will be cancelled (it uses switchMap by default).
+ * If the promise rejects, the error will be thrown.
+ *
+ * It can also be used with Observables:
+ *
+ * ```ts
+ * const value = computedAsync(() =>
+ *  this.http.get(`https://localhost/api/people/${this.userId()}`)
+ * );
+ * ```
+ *
+ * You can also pass an `initialValue` option to set the initial value of the computed value.
+ *
+ * ```ts
+ * const userTasks = computedAsync(() =>
+ *   this.http.get(`https://localhost/api/tasks?userId=${this.userId()}`),
+ *   { initialValue: [] }
+ * );
+ * ```
+ *
+ * You can also pass a `behavior` option to change the behavior of the computed value.
+ * - `switch` (default): will cancel the previous computation when a new one is triggered
+ * - `merge`: will use `mergeMap` to merge the last observable with the new one
+ * - `concat`: will use `concatMap` to concat the last observable with the new one
+ * - `exhaust`: will use `exhaustMap` to skip all the new emissions until the last observable completes
+ *
+ * You can also pass an `injector` option if you want to use it outside of the injection context.
+ *
+ * @param computation
+ * @param options
+ */
+export function computedAsync<T>(
+	computation: (previousValue?: T | undefined) => ComputationResult<T>,
+	options?: ComputedAsyncOptions<T>,
+): Signal<T> {
+	return assertInjector(computedAsync, options?.injector, () => {
+		const destroyRef = inject(DestroyRef);
+
+		// source$ is a Subject that will emit the new source value
+		const sourceEvent$ = new Subject<Promise<T> | Observable<T>>();
+
+		// will hold the current value
+		const sourceValue = signal<T | undefined>(
+			options?.initialValue ?? undefined,
+		);
+
+		const effectRef = effect(
+			() => {
+				// we need to have an untracked() here because we don't want to register the sourceValue as a dependency
+				// otherwise, we would have an infinite loop
+				const currentValue = untracked(() => sourceValue());
+
+				const newSource = computation(currentValue);
+				if (!isObservable(newSource) && !isPromise(newSource)) {
+					// if the new source is not an observable or a promise, we set the value immediately
+					untracked(() => sourceValue.set(newSource));
+					return;
+				}
+
+				// we untrack the source$.next() so that we don't register other signals as dependencies
+				untracked(() => sourceEvent$.next(newSource));
+			},
+			{ injector: options?.injector },
+		);
+
+		const source$: Observable<T> = createFlattenObservable(
+			sourceEvent$,
+			options?.behavior ?? 'switch',
+		);
+
+		const sourceResult = source$.subscribe({
+			next: (value) => sourceValue.set(value),
+			error: (error) => {
+				// NOTE: Error should be handled by the user (using catchError or .catch())
+				sourceValue.set(error);
+				throw error;
```

Comment:
There is no need to throw the error here, it is just an error notification. The error is not caught.